neo/Consensus/ConsensusService.cs

```diff
@@ -225,9 +286,281 @@ private void OnPersistCompleted(Block block)
+                    if (!regeneratedChangeView.Verify(snap) || !PerformBasicConsensusPayloadPreChecks(regeneratedChangeView)) continue;
+                    verifiedChangeViewWitnessInvocationScripts[i] = message.ChangeViewWitnessInvocationScripts[i];
+                    validChangeViewCount++;
+                    if (validChangeViewCount < context.M) continue;
+                    canRestoreView = true;
+                    break;
+                }
+            }
+
+            if (!canRestoreView) return;
+            // We had enough valid change view messages or preparations to safely change the view number.
+            Log($"regenerating view: {message.ViewNumber}");
+
+            if (context.PrimaryIndex == context.MyIndex && prepareRequestPayload != null)
+            {
+                // If we are the primary and there was a valid prepare request payload, we will accept our own prepare
+                // request, so avoid the normal IntializeConsensus behavior.
+                context.Reset(message.ViewNumber);
+                // Since we won't want to fill the context, we will behave like a backup even though we have the primary
+                // index; we set the backup state so we can call OnPrepareRequestReceived below.
+                context.State |= ConsensusState.Backup;
+                // We set the timer in the same way a Backup sets their timer in this case.
+                Log($"initialize: height={context.BlockIndex} view={message.ViewNumber} index={context.MyIndex} role={ConsensusState.Primary}");
+                ChangeTimer(TimeSpan.FromSeconds(Blockchain.SecondsPerBlock << (message.ViewNumber + 1)));
+            }
+            else
+            {
+                if (block_received_time == null)
+                {
+                    block_received_time = TimeProvider.Current.UtcNow - TimeSpan.FromSeconds(
+                                              Blockchain.TimePerBlock.TotalSeconds * Math.Pow(2, message.ViewNumber));
+                }
+                InitializeConsensus(message.ViewNumber);
+            }
+
+            for (int i = 0; i < context.Validators.Length; i++)
+            {
+                if (verifiedChangeViewWitnessInvocationScripts[i] != null)
+                {
+                    context.ChangeViewWitnessInvocationScripts[i] = verifiedChangeViewWitnessInvocationScripts[i];
+                    context.ChangeViewTimestamps[i] = message.ChangeViewTimestamps[i];
+                    context.OriginalChangeViewNumbers[i] = message.OriginalChangeViewNumbers[i];
+                    context.ExpectedView[i] = message.ViewNumber;
+                }
+            }
+
+            if (prepareRequestPayload != null)
+            {
+                Log($"regenerating prepare request");
+                // Note: If our node is the primary this will accept its own previously sent prepare request here.
+                OnPrepareRequestReceived(prepareRequestPayload, prepareRequest);
```

Comment:
Here, I would prefer that we set an additional flag for `OnPrepareRequestReceived`, `RegeneratingPrimary  = true`.
With that we gonna be more sure about what is going on and remove that flag I mentioned in the previous comment.

Reply:
Hmm.. I'll make some local changes and see if it looks cleaner. I'm not sure until I take a look at it, but there is also logic that happens in OnTimer that will probably have to change if I were to change it.

Reply:
If you would prefer it have both flags set, I will adjust it that way, since I don't see any harm in having them both set as long as it sets `RequestSent` again.

Reply:
I prefer both flags, it more clear. It is a little strange to set Backup flag for the Primary.

Reply:
Ok, I went ahead and added it to have both flags again, so I'm going to resolve it now. 

I think since it is the Primary, it already has sent the request, and it is going to accept it's own `PrepareRequest` it makes sense for it to have `Primary`, `RequestSent` and `Backup` flags.